src/main/java/com/comcast/zucchini/FlexibleBarrier.java

```diff
@@ -0,0 +1,234 @@
+
+        long milli = getMonoMilliseconds() + milliseconds;
+
+        if(milliseconds < 0)
+            this.primary.awaitAdvance(phase);
+        else {
+            //if it's getting interrupted exceptions and it hasn't actually timedout, then ignore them
+            while(true) {
+                try {
+                    this.primary.awaitAdvanceInterruptibly(
+                            phase,
+                            (milli - getMonoMilliseconds()),
+                            TimeUnit.MILLISECONDS);
+                    break;
+                }
+                catch(InterruptedException iex) {
+                    if(timedout)
+                        break;
+                }
+                catch(TimeoutException tex) {
+                    if(!timedout) {
+                        synchronized(this) {
+                            if(!timedout) {
+                                timedout = true;
+
+                                this.unlock();
+                            }
+                        }
+                    }
+                    break;
+                }
+            }
+        }
+
+        int ret = this.arrivePrimary();
+
+        //last one to release does the reset
+        if(1 == (this.arrivedThreads.size() - ret)) {
+            this.secondaryOrder = 0;
+            this.unlock();
+            this.timedout = false;
+        }
+
+        //secondary barrier to prevent overrun
+        this.secondary.arriveAndAwaitAdvance();
+
+        if(0 == this.arriveSecondary()) {
+            this.primaryOrder = 0;
+        }
+
+        LOGGER.trace("free {} as order {}", name(), ret);
+
+        return ret;
+    }
+
+    /**
+     * Decrements the number of parties that the current barrier is waiting for, and decreases the future number of parties to wait for.
+     *
+     * This is indicative of a party (Context) having crashed on the scenario and being irrecoverable.
+     *
+     * @author Andrew Benton
+     */
+    void dec() {
+        this.primary.arriveAndDeregister();
+        this.secondary.arriveAndDeregister();
+    }
+
+    /**
+     * Reset the FlexibleBarrier for the next intercept in the scenario.
+     *
+     * @author Andrew Benton
+     */
+    synchronized void reset() {
+        this.arrivedThreads.clear();
+        this.timedout = false;
+        this.primaryOrder = 0;
+        this.secondaryOrder = 0;
+    }
+
+    /**
+     * Reset the number of parties for the barrier back to full and removes arrived threads.
+     *
+     * @author Andrew Benton
+     */
+    synchronized void refresh() {
+        this.reset();
+        this.primary.bulkRegister(this.azt.contexts.size() - this.primary.getRegisteredParties());
+        this.secondary.bulkRegister(this.azt.contexts.size() - this.secondary.getRegisteredParties());
+    }
+
+    /**
+     * A shortcut to get the {@see TestContext} name or provide "<NULL>" it its place.
+     */
+    private static String name() {
```

Comment:
Factor this out to a utility class as it is used / defined in multiple places. You can use `import static` in your other classes to save on typing.